namui-cli/src/services/rust_build_service.rs

```diff
@@ -0,0 +1,266 @@
+use crate::{debug_println, types::ErrorMessage};
+use cargo_metadata::{diagnostic::DiagnosticLevel, CompilerMessage, Message};
+use std::{
+    io::Read,
+    path::PathBuf,
+    process::{Child, Command, Stdio},
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc, Mutex,
+    },
+    thread,
+    time::Duration,
+};
+
+pub struct RustBuildService {
+    builder: Mutex<Option<Arc<CancelableBuilder>>>,
+    is_build_just_started: AtomicBool,
+}
+pub enum BuildResult {
+    Canceled,
+    Successful(CargoBuildResult),
+    Failed(String), // It's not failed if cargo build result has error messages.
+}
+
+#[derive(Clone)]
+pub struct BuildOption {
+    pub dist_path: PathBuf,
+    pub project_root_path: PathBuf,
+    pub platform: BuildPlatform,
+}
+
+#[derive(Clone)]
+pub enum BuildPlatform {
+    WasmWeb,
+}
+
+impl RustBuildService {
+    pub(crate) fn new() -> Self {
+        Self {
+            builder: Mutex::new(None),
+            is_build_just_started: AtomicBool::new(false),
+        }
+    }
+
+    pub(crate) fn cancel_and_start_build(&self, build_option: &BuildOption) -> BuildResult {
+        if self.is_build_just_started.swap(true, Ordering::Relaxed) {
+            return BuildResult::Canceled;
+        }
+
+        let result_receiver = {
+            let mut builder_lock = self.builder.lock().unwrap();
+            if let Some(builder) = builder_lock.take() {
+                builder.cancel();
+            }
+
+            self.is_build_just_started.store(false, Ordering::Relaxed);
```

Comment:
Only one access can cancel the build, and during that time, other accesses will be canceled before trying to cancel the build.